staticd/static_bfd.c

```diff
@@ -0,0 +1,789 @@
+{
+	struct ipaddr ia_src = {}, ia_dst = {};
+	struct in6_addr *ia_srcp = NULL;
+
+	if (srg->srg_bsp == NULL)
+		return;
+
+	if (yang_dnode_exists(dnode, "../source")) {
+		yang_dnode_get_ip(&ia_src, dnode, "../source", NULL);
+		ia_srcp = (struct in6_addr *)&ia_src.ip;
+	} else
+		ia_srcp = NULL;
+
+	yang_dnode_get_ip(&ia_dst, dnode, "../peer", NULL);
+
+	if (ia_dst.ipa_type == IPADDR_V4)
+		bfd_sess_set_ipv4_addrs(srg->srg_bsp, (struct in_addr *)ia_srcp,
+					(struct in_addr *)&ia_dst.ip);
+	else
+		bfd_sess_set_ipv6_addrs(srg->srg_bsp,
+					(struct in6_addr *)ia_srcp,
+					(struct in6_addr *)&ia_dst.ip);
+
+	bfd_sess_install(srg->srg_bsp);
+}
+
+void static_route_group_bfd_interface(struct static_route_group *srg,
+				      const char *ifname)
+{
+	if (srg->srg_bsp == NULL)
+		return;
+
+	bfd_sess_set_interface(srg->srg_bsp, ifname);
+	bfd_sess_install(srg->srg_bsp);
+}
+
+static void static_route_group_bfd_updatecb(
+	__attribute__((unused)) struct bfd_session_params *bsp,
+	const struct bfd_session_status *bss, void *arg)
+{
+	struct static_route_group *srg = arg;
+	struct static_group_member *sgm;
+
+	TAILQ_FOREACH (sgm, &srg->srg_sgmlist, sgm_entry)
+		static_next_hop_bfd_change(sgm->sgm_sn, bss);
+}
+
+void static_route_group_bfd_enable(struct static_route_group *srg,
+				   const struct lyd_node *dnode)
+{
+	bool use_interface = yang_dnode_exists(dnode, "../interface");
+	bool use_profile = yang_dnode_exists(dnode, "../profile");
+	bool mhop = yang_dnode_get_bool(dnode, "../multi-hop");
+
+	/* Reconfigure or allocate new memory. */
+	if (srg->srg_bsp == NULL)
+		srg->srg_bsp =
+			bfd_sess_new(static_route_group_bfd_updatecb, srg);
+
+	static_route_group_bfd_addresses(srg, dnode);
+	bfd_sess_set_interface(
+		srg->srg_bsp,
+		use_interface ? yang_dnode_get_string(dnode, "../interface")
+			      : NULL);
+	bfd_sess_set_profile(srg->srg_bsp, use_profile ? yang_dnode_get_string(
+						   dnode, "../profile")
+						       : NULL);
+
+	bfd_sess_set_mininum_ttl(srg->srg_bsp, mhop ? BFD_MULTI_HOP_MIN_TTL
+						    : BFD_SINGLE_HOP_TTL);
+
+	/* Install or update the session. */
+	bfd_sess_install(srg->srg_bsp);
+}
+
+void static_route_group_bfd_multi_hop(struct static_route_group *srg, bool mhop)
+{
+	if (srg->srg_bsp == NULL)
+		return;
+
+	bfd_sess_set_mininum_ttl(srg->srg_bsp, mhop ? BFD_MULTI_HOP_MIN_TTL
+						    : BFD_SINGLE_HOP_TTL);
+	bfd_sess_install(srg->srg_bsp);
+}
+
+void static_route_group_bfd_disable(struct static_route_group *srg)
+{
+	bfd_sess_free(&srg->srg_bsp);
+}
+
+void static_route_group_bfd_profile(struct static_route_group *srg,
+				    const char *profile)
+{
+	if (srg->srg_bsp == NULL)
+		return;
+
+	bfd_sess_set_profile(srg->srg_bsp, profile);
+	bfd_sess_install(srg->srg_bsp);
+}
+
+/*
+ * Misc.
+ */
+static void
+static_route_group_var(vector comps,
+		       __attribute__((unused)) struct cmd_token *token)
+{
+	struct static_route_group *srg;
+
+	TAILQ_FOREACH (srg, &sbglobal.sbg_srglist, srg_entry)
```

Comment:
I think it should be implemented using NB. When we centralize the FRR management in a separate daemon, we won't have access to the internal daemon state so we'll need to implement the autocomplete through NB. Here we can easily implement it using iteration through NB config.